### PR TITLE
Fix peft inference

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?

1. Fix PEFT inference (bug due to trainer not attached during peft inference)
2. Add CI test for PEFT inference
3. Expose top_k to enable greedy sampling in inference script

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/peft.py [hunk 1]
```diff
@@ -105,7 +106,7 @@ def __call__(self, model: nn.Module) -> nn.Module:
         else:
             model.walk(self.transform)
 
-        if hasattr(model, "trainer") and model.trainer.state.fn != TrainerFn.FITTING:
+        if is_trainer_attached(model) and model.trainer.state.fn != TrainerFn.FITTING:
```
Comment (cuichenx) [inline nemo/lightning/pytorch/callbacks/peft.py:109]: this is the main change from this PR

**Review by HuiyingLi — DISMISSED** [review]
LGTM!

File: .github/workflows/cicd-main.yml [hunk 2]
```diff
@@ -4665,9 +4665,26 @@ jobs:
...
-        --lora_checkpoint_path=/home/TestData/nemo2_ckpt/llama_lora_ci_checkpoint/ \
+        --lora_checkpoint_path=/home/TestData/nemo2_ckpt/llama_lora_ci_checkpoint_v2/ \
         --output_path=/tmp/nemo2_lora_merge/${{ github.run_id }}
 
+  L2_NEMO_2_LoRA_Inference:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NEMO_2_LoRA_Inference') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:4676]: Please use the single GPU machine instead:

```suggestion
      RUNNER: self-hosted-azure-gpus-1
```

Also, make sure to add this test to the final check in `Nemo_CICD_Test`

**Review by ko3n1g — CHANGES_REQUESTED** [review]

File: .github/workflows/cicd-main.yml (hunk 2 above)
Comment (cuichenx) [reply .github/workflows/cicd-main.yml:4676]: Ah, thanks!

**Review by ko3n1g — APPROVED** [review]
